railties/lib/rails/commands/restart/restart_command.rb

```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Rails
+  module Command
+    class RestartCommand < Base # :nodoc:
+      desc "restart", "Restart app by touching tmp/restart.txt"
+      def perform
+        require "fileutils"
+        FileUtils.mkdir_p Rails::Command.root.join("tmp")
+        FileUtils.touch   Rails::Command.root.join("tmp/restart.txt")
```

Comment:
I think `Rails::Command.root` is better than `Rails.root` for the reason you mentioned, but it might be worth considering how `Rails::Command.root` will behave for plugin dummy apps.  For a plugin, [`bin/rails` sets both `ENGINE_ROOT` and `APP_PATH`](https://github.com/rails/rails/blob/ef4841eac873ee20856944d129cdb14abf05bf5e/railties/lib/rails/generators/rails/plugin/templates/bin/rails.tt#L4-L8), and `Rails::Command.root` will [return the former](https://github.com/rails/rails/blob/ef4841eac873ee20856944d129cdb14abf05bf5e/railties/lib/rails/command.rb#L109-L110).

So, what if we factored out an `Rails::Command.application_root` method and used that instead?

```ruby
def root
  defined?(ENGINE_ROOT) ? Pathname(ENGINE_ROOT) : application_root
end

def application_root
  Pathname(File.expand_path("../..", APP_PATH)) if defined?(APP_PATH)
end
```

Reply:
I'm wondering if we need to handle the case where `application_root` is nil.

Reply:
Ah, nevermind. It always gets set in the `bin/rails` as Johnathon pointed out.

Reply:
I've pushed these changes.